From 3bc510635e4e03b8546c0edb88f5f5cff77724e0 Mon Sep 17 00:00:00 2001
From: Sven Grunewaldt <strayer@olle-orks.org>
Date: Thu, 31 Oct 2013 21:06:11 +0100
Subject: [PATCH] Handle --disable-plugins* correctly in configure.ac

---
 configure.ac | 24 ++++++++++++++++++------
 1 file changed, 18 insertions(+), 6 deletions(-)

diff --git a/configure.ac b/configure.ac
index 2c0e8c1..8560275 100644
--- a/configure.ac
+++ b/configure.ac
@@ -69,22 +69,34 @@ AC_ARG_ENABLE(debug,
 plugins=disabled
 AC_ARG_ENABLE(plugins,
 [AS_HELP_STRING(--enable-plugins,Enable support for plugins - experimental)],
-[CPPFLAGS="$CPPFLAGS -DPLUGINS=1"
- plugins=enabled])
+[
+	AS_IF([test "x$enableval" != "xno"], [
+		CPPFLAGS="$CPPFLAGS -DPLUGINS=1"
+		plugins=enabled
+	])
+])
 AM_CONDITIONAL(PLUGINS, test x$plugins = xenabled)
 
 relaxed_plugins_permissions=disabled
 AC_ARG_ENABLE(relaxed-plugins-permissions,
 [AS_HELP_STRING(--enable-relaxed-plugins-permissions,Allow loading plugins owned by other users)],
-[CPPFLAGS="$CPPFLAGS -DRELAXED_PLUGINS_PERMISSIONS=1"
- relaxed_plugins_permissions=enabled])
+[
+	AS_IF([test "x$enableval" != "xno"], [
+		CPPFLAGS="$CPPFLAGS -DRELAXED_PLUGINS_PERMISSIONS=1"
+		relaxed_plugins_permissions=enabled
+	])
+])
 AM_CONDITIONAL(RELAXED_PLUGINS_PERMISSIONS, test x$relaxed_plugins_permissions = xenabled)
 
 plugins_root=disabled
 AC_ARG_ENABLE(plugins-root,
 [AS_HELP_STRING(--enable-plugins-root,Only load plugins sitting in the default plugins directory)],
-[CPPFLAGS="$CPPFLAGS -DENABLE_PLUGINS_ROOT=1"
- plugins_root=enabled])
+[
+	AS_IF([test "x$enableval" != "xno"], [
+		CPPFLAGS="$CPPFLAGS -DENABLE_PLUGINS_ROOT=1"
+		plugins_root=enabled
+	])
+])
 AM_CONDITIONAL(ENABLE_PLUGINS_ROOT, test x$plugins_root = xenabled)
 
 AC_SUBST([MAINT])
-- 
1.8.4.1

